json/src/spec/builtin.rs

```diff
@@ -92,13 +75,35 @@ pub struct Builtin {
 	pub pricing: Pricing,
 	/// Activation block.
 	pub activate_at: Option<Uint>,
-	/// EIP 1108
-	pub eip1108_transition: Option<Uint>,
+}
+
+/// Builtin price
+#[derive(Debug, PartialEq, Deserialize, Clone)]
+#[serde(rename_all = "snake_case")]
+#[serde(deny_unknown_fields)]
+#[serde(untagged)]
+pub enum Pricing {
```

Comment:
What is the reason we can't make all chains be "Multi"? Like, what does having `Single`/`Multi` buy us? The only reason I can think of is backwards compatibility (which is a good reason). What would it take to make serde cope with both types and deserialize to a vec of `PricingWithActivation`?

Reply:
It is for backward compatibility only I would say.

I don't think it is possible to have both types and deserialize to `Vec<PricingWithActivation>` in a backward-compatible way. However, another option is to add another field `Vec<PricingWithActivation>` and deprecated the old one but I'm open to suggestions if you come with something better

Reply:
I was sort of thinking about something akin to [`HashOrMap`](https://github.com/paritytech/parity-ethereum/blob/d9201aa6f2a03471c88f8c5638a7abe93d0db57c/json/src/spec/state.rs#L27-L38) that I used to deal with json like `"postState": "blabla"` vs `"postState" : { "stuff": "here"}`.
Not sure it'd work here but the situation there was similar: support two different types in the input JSON.

Reply:
I don't understand what you are after here. 
Pricing is similar to `HashOrMap` in your example but with an additional Multi. If you are referring to `Multi<Vec<PricingAt>` could be replaced `Vec<PricingAt>` that should be possible.

Otherwise, explain :wave:

Reply:
> If you are referring to `Multi<Vec<PricingAt>` could be replaced `Vec<PricingAt>` that should be possible.

Yep, that's it: avoid having the `Single`/`Multi` distinction in the code and instead deal with input JSON being effectively different for the single/multi cases when deserializing directly. Worth doing here in your opinion?

Reply:
all right, I see. Yes, it is worth doing

Reply:
well, I think I got confused by this again after investigated it or understood it wrong again sorry. 

The problem is that current chain spec has the following format for pricing:

```bash
// legacy
"pricing": { "linear": {"base": 60, "word": 12 }, "activate_at": 87}
// new
"pricing": [ ],
```
Having it deserialize to a `Vec` directly can't be done because it is the same type currently. Thus, `HashOrMap` example doesn't help because you still need to match on it to figure out whether it is `Array<Vec<T>>` or `Object(T)` at runtime.

As I see it we could do:
1. `provide a method `to_vec` which does the matching and then returns `Vec<PricingAt>`
2. create a new type `BuiltinV2` which just has an inner type `Vec/BTreeMap<PricingAt>`.

Have I got anything wrong or other suggestions?

Reply:
> Have I got anything wrong or other suggestions?

Most likely I made it simpler than it is. Let me check out the code and play with it.

Reply:
So, a tricky one for sure.
What I had in mind was having `pub struct Pricing(Vec<PricingAt>)` and then have a `#[serde(deserialize_with = "bla")]`. In `bla()` we'd check if it's a JSON `Array` or `Object` (or garbage, i.e. error) and the deserialize the right type:

 - if array, then assume it's a vec of `PricingAt`
 - if object, then check if it's a legacy `PriceInner` and if yes, deserialize it and build a `PriceAt` out of it; if no, deserialize a single `PricingAt`. Wrap in a `Vec` and return.

The point of it all is to move the messy input format handling into one spot and let the rest of the Rust code only ever see `Vec<PriceAt>`.

My serde is weak so I'm not sure how deep of a rabbit hole this is. I had a quick stab at it and failed so…

Reply:
Ok, thanks for the explanation. I can give it a try, would be nice to include to fail empty `Vec` as then I guess

Reply:
I didn't get it to work, see https://github.com/niklasad1/serde-manual-deserialize/blob/master/src/lib.rs#L74-#L120

So, I created a `compatibility layer` instead...